### PR TITLE
ScriptFileWatcher fixes for entire directories

### PR DESCRIPTION
**Pull request by ccutrer**

### CODE REVIEW

**Comment by J-N-K** [comment]
Can you add a short description what issue you are trying to solve? Thanks.

**Comment by ccutrer** [comment]
This fixes one obvious bug, and improves one thing that could be argued if it should be a bug or a feature:
1. If you add or remove an entire directory from the watched folder (by moving it in and out), it would not load all the new scripts/unload all the scripts that are now gone
2. If you add an entire directory at once, make sure to sort all of the new files together so that start levels are respected. This also applies to the initial load of all scripts if the file watcher is restarted (most commonly when you install or update the jsscripting or jrubyscripting addons).

**Review by J-N-K — APPROVED** [review]
LGTM, but since we are close to a release and this is very essential, I would prefer to have someone looking into it.

**Comment by ccutrer** [comment]
@florian-h05: any chance you could try this out for me, and confirm it doesn't cause any problems for JS?

**Review by florian-h05 — APPROVED** [review]
I haven't looked at the code, but I have tested this for JS Scripting, **it works fine!**

See my comment above for more information.

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java [hunk 1]
```diff
@@ -190,13 +226,24 @@ protected void processWatchEvent(@Nullable WatchEvent<?> event, @Nullable Kind<?
...
+                            }
+                        }
+                    } else {
+                        removeFile(new ScriptFileReference(file.toURI().toURL()));
+                    }
                 }
 
                 if (file.canRead() && (ENTRY_CREATE.equals(kind) || ENTRY_MODIFY.equals(kind))) {
-                    importFileWhenReady(new ScriptFileReference(fileUrl));
+                    importResources(file);
```
Comment (J-N-K) [inline bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java:246]:
```suggestion
                    collectResources(file).forEach(this::importFileWhenReady);
```

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java [hunk 2]
```diff
@@ -138,43 +140,77 @@ private void importInitialResources(File rootDirectory) {
...
             File[] files = rootDirectory.listFiles();
             if (files != null) {
+                Collection<ScriptFileReference> resources = new TreeSet<>();
                 for (File f : files) {
                     if (!f.isHidden()) {
-                        importResources(f);
+                        resources.addAll(collectResources(f));
                     }
                 }
+                importResources(resources);
```
Comment (J-N-K) [inline bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java:149]:
```suggestion
                resources.forEach(this::importFileWhenReady);
```

**Comment by J-N-K** [comment]
If it is confirmed working, I would merge if this is addressed.

**Review by J-N-K — APPROVED** [review]

**Comment by J-N-K** [comment]
Thanks.

**Pull request merged**